# Web Scrapping de Yuhmak
#### Autor: Ciriondev
#### Este documento intenta mostrar el alcance del motor de webscraping desarrollado por la empresa, para ello se utilizará la página pública de Yuhmak para extraer las motocicletas catalogadas junto con los atributos descriptivos de cada publicación.
#### Ago 08, 2024

### Extracción de datos

In [6]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
import pandas as pd
import time
from selenium.common.exceptions import NoSuchElementException

if 'google.colab' in str(get_ipython()):
    options = webdriver.ChromeOptions()
    options.add_argument('-headless')
    options.add_argument('-no-sandbox')
    options.add_argument('-disable-dev-shm-usage')

    driver = webdriver.Chrome('chromedriver',options=options)

else:
    service = Service(executable_path=r'/usr/local/bin/chromedriver')
    options = webdriver.ChromeOptions()
    options.add_argument('-headless')
    options.add_argument('-no-sandbox')
    options.add_argument('-disable-dev-shm-usage')
    options.add_argument("--window-size=1920,1200")

    driver = webdriver.Chrome(service=service,options=options)

Definimos la URL de la sección **Motos**:

In [7]:
driver.get("https://www.yuhmak.com/seccion-motos/")

Obtenemos el listado de todas las publicaciones disponibles en la página principal de la sección:

In [8]:
while True:
    try:
        ver = driver.find_element(By.LINK_TEXT, 'Mostrar más')
        ver.click()
        time.sleep(3)

    except NoSuchElementException as e:
        print(f"No se encontró elemento 'Mostrar más'")
        break

No se encontró elemento 'Mostrar más'


Encontramos los links a las publicaciones:

In [9]:
lista = driver.find_elements(By.CSS_SELECTOR, '.vtex-search-result-3-x-galleryItem.vtex-search-result-3-x-galleryItem--normal.vtex-search-result-3-x-galleryItem--grid.pa4')

In [10]:
links = []

for i in range(len(lista)):
    href = lista[i].find_element(By.CSS_SELECTOR, '.vtex-product-summary-2-x-clearLink.vtex-product-summary-2-x-clearLink--product-slider.h-100.flex.flex-column')
    links.append(href.get_attribute('href'))

In [11]:
len(links)

132

Procedemos a la recolección de los datos y su correspondiente análisis:

In [12]:
motos = pd.DataFrame()

for i in range(len(links) - 1):
        driver.get(links[i])
                
        try:
            titulo = driver.find_element(By.CLASS_NAME, 'yuhmak-yuhmak-installments-calculator-custom-0-x-productBrand ')
            tabla = driver.find_elements(By.CLASS_NAME, 'vtex-flex-layout-0-x-flexRow--productSpecificationItem')

            moto = dict({'Nombre': titulo.text})

            for i in range(len(tabla)):
                linea = tabla[i].text.split('\n')
                if len(linea) > 1:
                    atrib = linea[0]
                    valor = linea[1]
                    moto[atrib] = valor

            fila = pd.DataFrame([moto])
            motos = pd.concat([motos, fila], ignore_index=True)

        except NoSuchElementException as e:
            print(f"Error recolectando datos")

Error recolectando datos


Visualizamos la estructura de los datos obtenidos de manera dinámica:

In [19]:
motos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 33 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Nombre                    130 non-null    object
 1   Modelo                    57 non-null     object
 2   Peso                      79 non-null     object
 3   Tipo de moto              111 non-null    object
 4   Llantas                   64 non-null     object
 5   Transmisión               108 non-null    object
 6   Capacidad de combustible  106 non-null    object
 7   Suspensión delantera      102 non-null    object
 8   Suspensión trasera        102 non-null    object
 9   Freno delantero           111 non-null    object
 10  Freno trasero             111 non-null    object
 11  Neumático delantero       91 non-null     object
 12  Neumático trasero         91 non-null     object
 13  Tipo de motor             108 non-null    object
 14  Largo de moto             

Veamos la forma de nuestro dataset generado mediante webscraping:

In [14]:
motos.shape

(130, 33)

Vemos que contamos con 130 elementos con 33 características cada uno.

Quitamos las columnas de nombre 'Bloque *' para facilitar la visualización de otros atributos:

In [24]:
pattern_to_remove = 'Bloque'

filtered_columns = [col for col in motos.columns if pattern_to_remove not in col]

In [25]:
motos = motos[filtered_columns]

Veamos un resúmen de los primeros 50 elementos recolectados

In [37]:
motos.head(50)

Nombre  \
0                  Moto Honda Wave 110 S - 2024   
1          Moto Honda Wave 110 Cast Disk - 2024   
2          Moto Honda Xr 250 New Tornado - 2024   
3               Moto Bajaj Rouser 200 Ns - 2023   
4           Moto Motomel Blitz 110 (R/T) - 2022   
5                    Moto Honda Xr 150 L - 2024   
6   Moto Motomel Skua 150 Silver Edition - 2022   
7             Moto Gilera Smash Base 110 - 2023   
8                     Moto Benelli 180 S - 2023   
9                   Moto Siam Qu 110 R/T - 2023   
10                   Moto Honda Xr 190 L - 2024   
11      Moto Bajaj Dominar 400 Ug Tourer - 2024   
12    Moto Suzuki Ax 100 Motor 2 Tiempos - 2023   
13       Moto Gilera Smash Tunning 110 R - 2024   
14         Moto Corven Triax 250 Touring - 2022   
15               Moto Gilera Gx1 125 Std - 2022   
16        Moto Motomel Cg 150 S2 F/Disco - 2023   
17           Moto Bajaj Rouser Ns 160 Td - 2024   
18           Moto Motomel Dlx 110 Deluxe - 2024   
19               Moto Bajaj Rouser N 250 - 2024   
20                    Moto Suzuki Ax 100 - 2024   
21                 Moto Moto Siam Qu 110 - 2024   
22           Moto Corven Triax 150 (R/D) - 2022   
23            Moto Gilera Smash Very 110 - 2024   
24                 Moto Kawasaki Klx 300 - 2024   
25               Moto Gilera Piccola 150 - 2021   
26              Moto Corven Terrain 250X - 2020   
27              Moto Motomel Skua 150 V6 - 2024   
28                   Moto Benelli Tnt 15 - 2024   
29             Moto Bajaj Dominar 250 Ug - 2024   
30           Moto Corven Energy 110 Base - 2024   
31              Moto Zanella Rx 150 Z7 F - 2021   
32               Moto Kawasaki Ninja 400 - 2023   
33                   Moto Zanella Zr 150 - 2022   
34     Moto Zanella Styler Exclusive 150 - 2021   
35       Moto Motomel Cg 150 S2 Base 1.8 - 2024   
36            Moto Honda Cb 125F Twister - 2024   
37                        Moto Cf 400 Nk - 2024   
38              Moto Benelli Trk 251 Abs - 2023   
39           Moto Zanella Zb 110 Z1 Base - 2024   
40      Moto Zanella Patagonia Eagle 250 - 2022   
41           Moto Kawasaki Ninja 400 Abs - 2024   
42                 Moto Gilera Sahel 150 - 2025   
43           Gilera Smash 110 Full R New - 2022   
44               Hero X Pulse 200 On Off - 2023   
45               Moto Zanella Due 110 Rt - 2023   
46                    Moto Suzuki Gn 125 - 2023   
47                     Moto Voge 650 Dsx - 2024   
48                  Moto Motomel Xmm 250 - 2024   
49       Moto Moto Mondial Ld 110 Max Rt - 2025   

                             Modelo      Peso  Tipo de moto   Llantas  \
0           Honda Wave 110 S - 2024    101 kg           Cub     Rayos   
1                               NaN       NaN           NaN       NaN   
2   Honda Xr 250 New Tornado - 2024    150 kg        ON-OFF     Rayos   
3                               NaN    145 Kg         Naked  Aleación   
4                               NaN     99 KG           Cub     Rayos   
5             Honda Xr 150 L - 2024    150 kg        ON-OFF     Rayos   
6                              2022    123 Kg        ON-OFF     Rayos   
7                               NaN    100 kg           Cub     Rayos   
8                               NaN    147 kg         Naked  Aleación   
9                               NaN       NaN           NaN       NaN   
10            Honda Xr 190 L - 2024    133 kg        ON-OFF     Rayos   
11                              NaN       NaN       Touring       NaN   
12                              NaN     82 kg           NaN     Rayos   
13                    Tunning 110 R    100 Kg           Cub       NaN   
14                Triax 250 Touring    152 kg       Touring       NaN   
15                              NaN       NaN         Naked  Aleación   
16                              NaN    115 kg        ON-OFF     Rayos   
17                              NaN       NaN         Naked       NaN   
18                              NaN       NaN       